File: wheel_of_fortune/schemas.py
from enum import Enum
from pydantic import BaseModel, Field


# -----------------------------------------------------------------------------
# Encoder
# -----------------------------------------------------------------------------


class EncoderState(BaseModel):
    sector: int
    rpm: float
    total_revs: float
    total_sectors: int
    missed_sector_count: int
    standstill: bool


class EncoderTestParams(BaseModel):
    initial_speed: float = Field(ge=0.0, le=100, examples=[20])
    drag_factor: float = Field(ge=0.0, le=100, examples=[0.1])


# -----------------------------------------------------------------------------
# Servos
# -----------------------------------------------------------------------------


class ServoState(BaseModel):
    pos: float
    duty: float
    detached: bool


class ServoStateIn(BaseModel):
    pos: float | None = Field(default=None, ge=-0.3, le=1.3, examples=[0.5])
    detached: bool | None = None


class ServoInfo(BaseModel):
    mount_angle: float
    zero_duty: float
    full_duty: float
    mount_duty: float
    mount_pos: float


class ServosState(BaseModel):
    motors: dict[str, ServoState]


class ServosStateIn(BaseModel):
    motors: dict[str, ServoStateIn] = {}


class ServosInfo(BaseModel):
    version: str
    motors: dict[str, ServoInfo]


# -----------------------------------------------------------------------------
# LEDs
# -----------------------------------------------------------------------------


class LedSegmentState(BaseModel):
    enabled: bool
    brightness: float
    palette: str
    primary_color: str
    secondary_color: str
    effect: str
    effect_speed: float
    effect_intensity: float


class LedSegmentStateIn(BaseModel):
    enabled: bool | None = True
    brightness: float | None = Field(default=1.0, ge=0.0, le=1.0)
    palette: str | None = "default"
    primary_color: str | None = "#FF0000"
    secondary_color: str | None = "#000000"
    effect: str | None = "solid"
    effect_speed: float | None = Field(default=0.5, ge=0.0, le=1.0)
    effect_intensity: float | None = Field(default=0.5, ge=0.0, le=1.0)


class LedsState(BaseModel):
    power_on: bool
    brightness: float
    segments: dict[str, LedSegmentState]


class LedsStateIn(BaseModel):
    brightness: float | None = Field(default=None, ge=0.0, le=1.0, examples=[0.5])
    segments: dict[str, LedSegmentStateIn] | None = None
    transition_ms: float = 0.0


class LedsInfo(BaseModel):
    version: str


# -----------------------------------------------------------------------------
# Sound system
# -----------------------------------------------------------------------------


class SoundChannelState(BaseModel):
    volume: float
    sound_name: str | None


class SoundChannelStateIn(BaseModel):
    volume: float | None = Field(default=None, ge=0.0, le=1.0, examples=[0.5])
    sound_name: str | None = None


class SoundSystemState(BaseModel):
    channels: dict[str, SoundChannelState]


class SoundSystemStateIn(BaseModel):
    channels: dict[str, SoundChannelStateIn] = {}


class SoundInfo(BaseModel):
    duration_secs: float


class SoundSystemInfo(BaseModel):
    sounds: dict[str, SoundInfo]


# -----------------------------------------------------------------------------
# Sectors
# -----------------------------------------------------------------------------


class SectorState(BaseModel):
    index: int
    name: str
    effect_id: str


class SectorStateIn(BaseModel):
    name: str | None = None
    effect_id: str | None = None


# -----------------------------------------------------------------------------
# Themes
# -----------------------------------------------------------------------------


class ThemeInfo(BaseModel):
    name: str
    description: str
    image_url: str
    based_on: list[str] = []
    startup_sound: str
    theme_sounds: list[str]
    poweroff_sound: str
    startup_led_preset: dict[str, LedSegmentStateIn]
    idle_led_preset: dict[str, LedSegmentStateIn]
    spinning_led_preset: dict[str, LedSegmentStateIn]
    standby_led_preset: dict[str, LedSegmentStateIn]
    poweroff_led_preset: dict[str, LedSegmentStateIn]


# -----------------------------------------------------------------------------
# Effects
# -----------------------------------------------------------------------------


class EffectInfo(BaseModel):
    name: str
    description: str
    image_url: str
    color: str
    based_on: list[str] = []
    effect_sound: str
    leds_preset: dict[str, LedSegmentStateIn]
    active_servos: list[str]


# -----------------------------------------------------------------------------
# Wheel
# -----------------------------------------------------------------------------


class WheelState(BaseModel):
    active_task: str
    theme_id: str
    standby_timer: float
    sectors: list[SectorState]
    encoder: EncoderState
    servos: ServosState
    leds: LedsState
    soundsystem: SoundSystemState


class WheelStateIn(BaseModel):
    active_task: str | None = None
    theme_id: str | None = None
    standby_timer: float | None = Field(default=None, examples=[600])
    sectors: dict[int, SectorStateIn] = {}
    servos: ServosStateIn | None = None
    leds: LedsStateIn | None = None
    soundsystem: SoundSystemStateIn | None = None


class WheelStateUpdate(BaseModel):
    active_task: str | None = None
    theme_id: str | None = None
    standby_timer: float | None = None
    sectors: list[SectorState] | None = None
    encoder: EncoderState | None = None
    servos: ServosState | None = None
    leds: LedsState | None = None
    soundsystem: SoundSystemState | None = None


class WheelInfo(BaseModel):
    version: str
    name: str
    display_name: str
    logo_name: str
    themes: dict[str, ThemeInfo]
    effects: dict[str, EffectInfo]
    servos: ServosInfo
    leds: LedsInfo
    soundsystem: SoundSystemInfo


# -----------------------------------------------------------------------------
# Websocket
# -----------------------------------------------------------------------------


class WsCommandType(Enum):
    INIT = "init"  # Full state and info (to client)
    UPDATE = "update"  # State update (to client)
    SET_STATE = "set_state"  # Set state (to server)


class WsInitPacket(BaseModel):
    cmd: WsCommandType = WsCommandType.INIT
    ts: float
    state: WheelState
    info: WheelInfo


class WsUpdatePacket(BaseModel):
    cmd: WsCommandType = WsCommandType.UPDATE
    ts: float
    update: WheelStateUpdate


class WsSetStatePacket(BaseModel):
    cmd: WsCommandType = WsCommandType.SET_STATE
    ts: float
    state: WheelStateIn

File: wheel_of_fortune/frontend/src/schemas.ts
import { z } from 'zod';

// ----------------------------------------------------------------------------
// Encoder
// ----------------------------------------------------------------------------

export const EncoderState = z.object({
  sector: z.number().int(),
  rpm: z.number(),
  total_revs: z.number(),
  total_sectors: z.number().int(),
  missed_sector_count: z.number().int(),
  standstill: z.boolean(),
});
export type EncoderState = z.infer<typeof EncoderState>;

// ----------------------------------------------------------------------------
// Servos
// ----------------------------------------------------------------------------

export const ServoState = z.object({
  pos: z.number(),
  duty: z.number(),
  detached: z.boolean(),
});
export type ServoState = z.infer<typeof ServoState>;

export const ServoStateIn = z.object({
  pos: z.number().optional(),
  detached: z.boolean().optional(),
});
export type ServoStateIn = z.infer<typeof ServoStateIn>;

export const ServoInfo = z.object({
  mount_angle: z.number(),
  zero_duty: z.number(),
  full_duty: z.number(),
  mount_duty: z.number(),
  mount_pos: z.number(),
});
export type ServoInfo = z.infer<typeof ServoInfo>;

export const ServosState = z.object({
  motors: z.record(z.string(), ServoState),
});
export type ServosState = z.infer<typeof ServosState>;

export const ServosStateIn = z.object({
  motors: z.record(z.string(), ServoStateIn).default({}),
});
export type ServosStateIn = z.infer<typeof ServosStateIn>;

export const ServosInfo = z.object({
  version: z.string(),
  motors: z.record(z.string(), ServoInfo),
});
export type ServosInfo = z.infer<typeof ServosInfo>;

// ----------------------------------------------------------------------------
// LEDs
// ----------------------------------------------------------------------------

export const LedSegmentState = z.object({
  enabled: z.boolean(),
  brightness: z.number(),
  palette: z.string(),
  primary_color: z.string(),
  secondary_color: z.string(),
  effect: z.string(),
  effect_speed: z.number(),
  effect_intensity: z.number(),
});
export type LedSegmentState = z.infer<typeof LedSegmentState>;

export const LedsState = z.object({
  power_on: z.boolean(),
  brightness: z.number(),
  segments: z.record(z.string(), LedSegmentState),
});
export type LedsState = z.infer<typeof LedsState>;

export const LedsStateIn = z.object({
  brightness: z.number().optional(),
  transition_ms: z.number().optional(),
});
export type LedsStateIn = z.infer<typeof LedsStateIn>;

export const LedsInfo = z.object({
  version: z.string(),
});
export type LedsInfo = z.infer<typeof LedsInfo>;

// ----------------------------------------------------------------------------
// Sound system
// ----------------------------------------------------------------------------

export const SoundInfo = z.object({
  duration_secs: z.number(),
});
export type SoundInfo = z.infer<typeof SoundInfo>;

export const SoundChannelState = z.object({
  volume: z.number(),
  sound_name: z.string().nullable().optional(),
});
export type SoundChannelState = z.infer<typeof SoundChannelState>;

export const SoundChannelStateIn = z.object({
  volume: z.number().optional(),
  sound_name: z.string().optional(),
});
export type SoundChannelStateIn = z.infer<typeof SoundChannelStateIn>;

export const SoundSystemState = z.object({
  channels: z.record(z.string(), SoundChannelState),
});
export type SoundSystemState = z.infer<typeof SoundSystemState>;

export const SoundSystemStateIn = z.object({
  channels: z.record(z.string(), SoundChannelStateIn).optional(),
});
export type SoundSystemStateIn = z.infer<typeof SoundSystemStateIn>;

export const SoundSystemInfo = z.object({
  sounds: z.record(z.string(), SoundInfo),
});
export type SoundSystemInfo = z.infer<typeof SoundSystemInfo>;

// ----------------------------------------------------------------------------
// Sectors
// ----------------------------------------------------------------------------

export const SectorState = z.object({
  index: z.number().int(),
  name: z.string(),
  effect_id: z.string(),
});
export type SectorState = z.infer<typeof SectorState>;

export const SectorStateIn = z.object({
  name: z.string().optional(),
  effect_id: z.string().optional(),
});
export type SectorStateIn = z.infer<typeof SectorStateIn>;

// ----------------------------------------------------------------------------
// Themes
// ----------------------------------------------------------------------------

export const ThemeInfo = z.object({
  name: z.string(),
  description: z.string(),
  image_url: z.string(),
  based_on: z.array(z.string()),
  startup_sound: z.string(),
  theme_sounds: z.array(z.string()),
  poweroff_sound: z.string(),
  // startup_led_preset: z.record(z.string(), LedSegmentStateIn),
  // idle_led_preset: z.record(z.string(), LedSegmentStateIn),
  // spinning_led_preset: z.record(z.string(), LedSegmentStateIn),
  // poweroff_led_preset: z.record(z.string(), LedSegmentStateIn),
});
export type ThemeInfo = z.infer<typeof ThemeInfo>;

// ----------------------------------------------------------------------------
// Effects
// ----------------------------------------------------------------------------

export const EffectInfo = z.object({
  name: z.string(),
  description: z.string(),
  image_url: z.string(),
  color: z.string(),
  based_on: z.array(z.string()),
  effect_sound: z.string(),
  // leds_preset: z.record(z.string(), LedSegmentStateIn),
  active_servos: z.array(z.string()),
});

export type EffectInfo = z.infer<typeof EffectInfo>;

// ----------------------------------------------------------------------------
// Wheel
// ----------------------------------------------------------------------------

export const WheelState = z.object({
  active_task: z.string().optional(),
  theme_id: z.string(),
  standby_timer: z.number(),
  sectors: z.array(SectorState),
  encoder: EncoderState,
  leds: LedsState,
  servos: ServosState,
  soundsystem: SoundSystemState,
});
export type WheelState = z.infer<typeof WheelState>;

export const WheelStateIn = z.object({
  theme_id: z.string().optional(),
  standby_timer: z.number().optional(),
  sectors: z.record(z.number().int(), SectorStateIn).optional(),
  leds: LedsStateIn.optional(),
  servos: ServosStateIn.optional(),
  soundsystem: SoundSystemStateIn.optional(),
});
export type WheelStateIn = z.infer<typeof WheelStateIn>;

export const WheelStateUpdate = z.object({
  active_task: z.string().optional(),
  theme_id: z.string().optional(),
  standby_timer: z.number().optional(),
  sectors: z.array(SectorState).optional(),
  encoder: EncoderState.optional(),
  servos: ServosState.optional(),
  leds: LedsState.optional(),
  soundsystem: SoundSystemState.optional(),
});
export type WheelStateUpdate = z.infer<typeof WheelStateUpdate>;

export const WheelInfo = z.object({
  version: z.string(),
  name: z.string(),
  display_name: z.string(),
  logo_url: z.string(),
  themes: z.record(z.string(), ThemeInfo),
  effects: z.record(z.string(), EffectInfo),
  servos: ServosInfo,
  leds: LedsInfo,
  soundsystem: SoundSystemInfo,
});
export type WheelInfo = z.infer<typeof WheelInfo>;

// ----------------------------------------------------------------------------
// Websocket
// ----------------------------------------------------------------------------

export const WsInitPacket = z.object({
  cmd: z.string(),
  ts: z.number(),
  state: WheelState,
  info: WheelInfo,
});
export type WsInitPacket = z.infer<typeof WsInitPacket>;

export const WsUpdatePacket = z.object({
  cmd: z.string(),
  ts: z.number(),
  update: WheelStateUpdate,
});
export type WsUpdatePacket = z.infer<typeof WsUpdatePacket>;

export const WsSetStatePacket = z.object({
  cmd: z.string(),
  ts: z.number(),
  state: WheelStateIn,
});
export type WsSetStatePacket = z.infer<typeof WsSetStatePacket>;
